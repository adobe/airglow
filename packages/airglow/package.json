{
  "name": "airglow",
  "version": "0.5.1",
  "repository": {
    "type": "git",
    "url": "https://github.com/adobe/airglow.git",
    "directory": "packages/airglow"
  },
  "publishConfig": {
    "access": "public"
  },
  "description": "Simple Clean React Framework",
  "author": "Adobe Systems",
  "license": "Apache-2.0",
  "files": [
    "dist/"
  ],
  "main": "dist/index.js",
  "peerDependencies": {
    "react": "^16.8.6"
  },
  "dependencies": {
    "ramda": "^0.26.1"
  },
  "devDependencies": {
    "@babel/cli": "^7.4.4"
  },
  "scripts": {
    "prebuild": "rm -rf dist && mkdir dist",
    "build": "babel src --out-dir dist --root-mode upward",
    "test": "echo \"Error: no test specified\" && exit 1"
  }
}
