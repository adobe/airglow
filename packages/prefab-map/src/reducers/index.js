/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { call } from '@airglow/reducers';
import * as T from '../actions';
import toggleModeReducer from './toggle.mode.reducer';
import { buttonZoomReducer, zoomReducer } from './zoom.reducer';
import positionReducer from './position.reducer';

export default call(
  call(buttonZoomReducer).for(T.BUTTON_ZOOM),
  call(zoomReducer).for(T.ZOOM),
  call(toggleModeReducer).for(T.TOGGLE_MODE),
  call(positionReducer).for(T.CHANGE_POSITION)
);
