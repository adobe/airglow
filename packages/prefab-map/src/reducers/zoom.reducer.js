/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { DEFAULT_ZOOM } from '../constants';
import clampZoom from '../util/clamp.map.zoom';

const constructPath = ['construct', 'defaultZoom'];
const zoomPath = ['store', 'zoom'];

const getZoom = state =>
  R.defaultTo(
    R.path(constructPath, state) || DEFAULT_ZOOM,
    R.path(zoomPath, state)
  );
const calcZoom = (current, size) => clampZoom(current + size);

export const buttonZoomReducer = (state, { payload }) =>
  R.assocPath(
    zoomPath,
    calcZoom(getZoom(state), payload.size),
    state
  );

export const zoomReducer = (state, { payload }) =>
  R.assocPath(
    zoomPath,
    payload.zoom,
    state
  );
