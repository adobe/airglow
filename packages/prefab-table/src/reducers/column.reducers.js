/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

import columnPrefab from '../column.prefab';

const addedColumnsPath = ['store', 'addedColumns'];

const hiddenColumnsPath = ['store', 'hiddenColumns'];

const addedColumnsLens = R.lensPath(addedColumnsPath);

const hiddenColumnsLens = R.lensPath(hiddenColumnsPath);

export const addColumnReducer = (state, { payload }) => {
  const { column, config } = payload;
  return R.over(addedColumnsLens, R.assoc(column, columnPrefab({ ...config })), state);
};

export const removeColumnReducer = (state, { payload }) => {
  const { column } = payload;
  return R.over(addedColumnsLens, R.dissoc(column), state);
};

export const showColumnReducer = (state, { payload }) => {
  const { column } = payload;
  return R.over(hiddenColumnsLens, R.without([column]), state);
};

export const hideColumnReducer = (state, { payload }) => {
  const { column } = payload;
  const hiddenColumns = R.pipe(
    R.path(hiddenColumnsPath),
    R.append(column),
    R.uniq
  )(state);

  return R.set(hiddenColumnsLens, hiddenColumns, state);
};
