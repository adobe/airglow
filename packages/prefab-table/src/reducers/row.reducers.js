/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

const allowsSelectionPath = ['construct', 'allowsSelection'];

const allowsMultipleSelectionPath = ['construct', 'allowsMultipleSelection'];

const selectedRowsPath = ['store', 'selectedRows'];

const selectedRowsLens = R.lensPath(selectedRowsPath);

export const deselectRowsReducer = (state, { payload }) => {
  const { rows } = payload;
  const normalized = !rows
    ? R.view(selectedRowsLens, state)
    : Array.isArray(rows) ? rows : [rows];

  return R.over(selectedRowsLens, R.without(normalized), state);
};

export const toggleRowReducer = (state, { payload }) => {
  if (!R.path(allowsSelectionPath, state)) return state;

  const { row } = payload;

  if (R.indexOf(row, R.pathOr([], selectedRowsPath, state)) >= 0) {
    return R.over(selectedRowsLens, R.without([row]), state);
  }

  return R.path(allowsMultipleSelectionPath, state)
    ? R.over(selectedRowsLens, R.append(row), state)
    : R.set(selectedRowsLens, [row], state);
};
