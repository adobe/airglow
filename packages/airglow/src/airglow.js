/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import React from 'react';
import { AirglowProvider, AirglowWrapper } from './airglow.context';
import Engine from './engine';
import { COMPONENT, HOC } from './types';

class Airglow extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.store = this.props.store ? this.props.store() : null;
    this.engine = new Engine(this.store, R.reverse(this.props.plugins || []));

    if (this.props.bootstrap) {
      this.engine.plugin(
        HOC,
        { component: AirglowWrapper, props: { config: this.props.bootstrap } }
      );
    }
  }

  renderChildren() {
    // take our children and merge them with the render plugins
    return R.reduce(
      (children, plugin) =>
        pluginReducers[plugin.type](plugin.callback, children),
      this.props.children,
      R.reverse(getRenderPlugins(this.engine))
    );
  }

  render() {
    const Renderer = this.props.renderer || AirglowProvider;
    return (
      <Renderer
        value={{
          bootstrap: (...args) => this.engine.bootstrapModule(...args),
          dispatch: action => this.engine.getStore().dispatch(action)
        }}
      >
        {this.renderChildren()}
      </Renderer>
    );
  }
}

const reduceHoc = ({ component: Component, props }, children) =>
  <Component {...props}>{children}</Component>;

const reduceComponent = (component, children) => (
  <div id="airglowRootWrapper">
    {children}
    {component}
  </div>
);

const pluginReducers = {
  [HOC]: reduceHoc,
  [COMPONENT]: reduceComponent
};

const getRenderPlugins = engine => R.reverse(R.sortBy(
  R.prop('index'),
  R.concat(
    engine.getPluginsData(HOC) || [],
    engine.getPluginsData(COMPONENT) || []
  )
));

export default Airglow;
