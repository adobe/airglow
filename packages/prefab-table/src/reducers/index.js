/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { call } from '@airglow/reducers';
import * as T from '../actions';
import { addColumnReducer, hideColumnReducer, removeColumnReducer, showColumnReducer } from './column.reducers';
import { deselectRowsReducer, toggleRowReducer } from './row.reducers';
import { sortReducer } from './sort.reducers';

export default call(
  call(addColumnReducer).for(T.ADD_COLUMN),
  call(deselectRowsReducer).for(T.DESELECT_ROWS),
  call(hideColumnReducer).for(T.HIDE_COLUMN),
  call(removeColumnReducer).for(T.REMOVE_COLUMN),
  call(showColumnReducer).for(T.SHOW_COLUMN),
  call(sortReducer).for(T.DO_SORT),
  call(toggleRowReducer).for(T.TOGGLE_ROW)
);
