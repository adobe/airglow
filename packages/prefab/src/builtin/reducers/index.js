/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/


import { call } from '@airglow/reducers';
import * as T from '../actions';

import {
  turnOnReducer, turnOffReducer, toggleReducer
} from './toggle.reducers';

import {
  valueReducer, resetReducer
} from './value.reducers';

import {
  listReducer, addToListReducer, removeFromListReducer, removeAtIndexReducer
} from './list.reducers';


export const reduceToggle = call(
  call(turnOnReducer).for(T.TOGGLE_ON),
  call(turnOffReducer).for(T.TOGGLE_OFF),
  call(toggleReducer).for(T.TOGGLE_SWITCH)
);

export const reduceValue = call(
  call(valueReducer).for(T.VALUE),
  call(resetReducer).for(T.VALUE_RESET)
);

export const reduceList = call(
  call(listReducer).for(T.LIST_SET),
  call(addToListReducer).for(T.LIST_ADD),
  call(removeFromListReducer).for(T.LIST_REMOVE),
  call(removeAtIndexReducer).for(T.LIST_REMOVE_AT)
);
