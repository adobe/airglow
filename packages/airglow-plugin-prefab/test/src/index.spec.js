/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { REDUCER, BOOTSTRAP_MODULE } from 'airglow';
import { reducer } from '@airglow/prefab';
import * as wrinkle from '@airglow/reducers';
import plugin from '../../src/index';

let engine;
let slice;
let dispatch;
const getCall = (type) => {
  for (let i = 0; i < engine.plugin.callCount; ++i) {
    if (engine.plugin.getCall(i).args[0] === type) {
      return engine.plugin.getCall(i).args[1];
    }
  }
  return null;
};

describe('AirglowPrefabPlugin', () => {
  beforeEach(() => {
    engine = { plugin: sinon.spy() };
    slice = { with: sinon.stub().returns('SLICED!') };
    sinon.stub(wrinkle, 'slice').returns(slice);
    dispatch = sinon.spy();

    plugin({ storeKey: 'myprefab' })(engine);
  });
  afterEach(() => {
    wrinkle.slice.restore();
  });

  it('should register reducers', function () {
    expect(getCall(REDUCER)).toBe('SLICED!');
    expect(slice.with.firstCall.args[0]).toBe(reducer);
  });

  it('should send the prefab content on bootstrap', function () {
    plugin()(engine);
    const bootstrap = getCall(BOOTSTRAP_MODULE);
    bootstrap({ dispatch, prefab: 'new.prefab.content' });
    expect(dispatch.lastCall.args).toMatchSnapshot();
  });

  it('doesnt bootstrap if null prefab', function () {
    plugin()(engine);
    const bootstrap = getCall(BOOTSTRAP_MODULE);
    bootstrap({ dispatch });
    expect(dispatch.callCount).toBe(0);
  });
});
