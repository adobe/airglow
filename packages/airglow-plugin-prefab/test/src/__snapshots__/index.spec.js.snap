// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirglowPrefabPlugin should send the prefab content on bootstrap 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": "new.prefab.content",
    "type": "BOOTSTRAP_PREFAB",
  },
]
`;
